<!DOCTYPE html>
<html lang="en-us">
    <head>
         
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>Time Course Analysis - Example workflow</title>
        
        <link rel="icon" href="/documentation/favicon.ico">
        
        <style>

    html body {
        font-family: '', sans-serif;
        background-color: white;
    }

    :root {
        --accent: #2196F3;
        --border-width:  0 ;
    }

</style>


<link rel="stylesheet" href="/documentation/css/main.css">





<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=">





<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">


<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">
 






<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>


<script>$(document).on('click', function() { $('.collapse').collapse('hide'); })</script>
 <meta name="generator" content="Hugo 0.57.1" />
        

        

        
        
        <script type="text/javascript" async
          src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
        </script>
        
    </head>

    <body>
         
        <nav class="navbar navbar-default navbar-fixed-top">

            <div class="container">

                <div class="navbar-header">

                    <a class="navbar-brand visible-xs" href="#">Time Course Analysis - Example workflow</a>

                    <button class="navbar-toggle" data-target=".navbar-collapse" data-toggle="collapse">
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                    </button>

                </div>

                <div class="collapse navbar-collapse">

                    
                        <ul class="nav navbar-nav">
                            
                                <li><a href="/documentation/">Home</a></li>
                            
                                <li><a href="/documentation/about/">About</a></li>
                            
                                <li><a href="/documentation/post/">Posts</a></li>
                            
                        </ul>
                    

                    
                        <ul class="nav navbar-nav navbar-right">
                            
                                <li class="navbar-icon"><a href="https://github.com/nibsbioinformatics/core"><i class="fa fa-github"></i></a></li>
                            
                                <li class="navbar-icon"><a href="https://twitter.com/nibsc_mhra"><i class="fa fa-twitter"></i></a></li>
                            
                        </ul>
                    

                </div>

            </div>

        </nav>


<main>

    <div class="item">

    
    
    

    
    

    <h4><a href="/documentation/2019/09/05/time-course-analysis-example-workflow/">Time Course Analysis - Example workflow</a></h4>
    <h5>September 5, 2019</h5>
    
    <a href="/documentation/tags/rnaseq"><kbd class="item-tag">RNAseq</kbd></a>
    

</div>


    <br> <div class="text-justify">


<pre class="r"><code># setwd(&quot;~/PROJECTS_WORK/projectnumber&quot;)
library(knitr)
opts_chunk$set(message=FALSE, error=FALSE, warning=FALSE, tidy=TRUE, tidy.opts=list(blank=FALSE), cache=TRUE, cache.lazy = FALSE, echo = TRUE,  results = &#39;asis&#39;, fig.pos=&#39;H&#39;, fig.wide=TRUE, eval=FALSE)
library(tidyverse)
library(reshape2)
library(pander)
library(grid)
library(VennDiagram)
library(DESeq2)
library(ImpulseDE2)
library(splines)
library(pheatmap)
library(TCseq)
library(biomaRt)
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(pander)
library(eulerr)
library(qvalue)
library(clusterProfiler)
library(org.Hs.eg.db)
library(DOSE)
# ddsHTSeq &lt;- readRDS(&quot;PXXX_ddsHTSeq_timeCourse_spline.RDS&quot;)
# sampleTable &lt;- readRDS(&quot;PXXX_sampleTable_metadata.RDS&quot;)
# objectImpulseDE2 &lt;- readRDS(&quot;PXXX_objectImpulseDE2.RDS&quot;)</code></pre>
<div id="preparing-data" class="section level1">
<h1>Preparing Data</h1>
<div id="loading-htseq-counts" class="section level2">
<h2>Loading HTSeq Counts</h2>
<p>This is run on the cluster, then an RDS file is analysed locally.
We build a metadata table for all samples, using the information coded in the sample names.</p>
<pre class="r"><code>### this is run on the cluster while the rest is run locally
directory &lt;- &quot;/usr/share/sequencing/projects/132/deseq&quot;
sampleFiles &lt;- grep(&quot;Bio&quot;, list.files(directory), value = TRUE)
### build metadata table starting from name -
### 132_Bio-1-D_wk-1_S15.htseq.a5.txt
sampleTable &lt;- data.frame(sampleName = sub(&quot;^132_(Bio-.+)-.{1}_wk-.{1}_S.+.txt$&quot;, 
    &quot;\\1&quot;, sampleFiles), fileName = sampleFiles, condition = ifelse(grepl(&quot;-D_&quot;, 
    sampleFiles), &quot;DMEM&quot;, &quot;XVIVO&quot;), time = sub(&quot;^132_Bio-.+-.{1}_wk-(.{1})_S.+.txt$&quot;, 
    &quot;\\1&quot;, sampleFiles), stringsAsFactors = FALSE)
## sample names need to be unique
sampleTable$sampleName &lt;- paste0(sampleTable$sampleName, &quot;_&quot;, sampleTable$condition, 
    &quot;_&quot;, sampleTable$time)
sampleTable &lt;- numerize(sampleTable, c(&quot;time&quot;))</code></pre>
</div>
<div id="modeling-time" class="section level2">
<h2>Modeling Time</h2>
<p>Following the suggestions from the Impulse2 paper, we use a spline function to model the time course. We create a natural cubic spline basis with 4 degrees of freedom on the data from the time column of the sample metadata.</p>
<pre class="r"><code>matTimeSplineBasis &lt;- ns(sampleTable$time, df = 4)
colnames(matTimeSplineBasis) &lt;- paste0(&quot;spline&quot;, seq(1, dim(matTimeSplineBasis)[2]))</code></pre>
<p>Then we embed the 4 splines into the sample metadata table we need to pass to DESeq2 in order to load the counts</p>
<pre class="r"><code>sampleTable &lt;- cbind(sampleTable, matTimeSplineBasis)</code></pre>
<p>And we create a design as a model with spline basis vectors as individual linear predictors (using the interaction terms rather than a stratified model only)</p>
<pre class="r"><code>ddsHTSeq &lt;- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable, directory = directory, 
    design = ~condition + condition:spline1 + condition:spline2 + condition:spline3 + 
        condition:spline4)</code></pre>
</div>
<div id="prefiltering-counts" class="section level2">
<h2>Prefiltering counts</h2>
<pre class="r"><code>keep &lt;- rowSums(counts(ddsHTSeq)) &gt;= 10
ddsHTSeq &lt;- ddsHTSeq[keep, ]
ddsHTSeq$condition &lt;- relevel(ddsHTSeq$condition, ref = &quot;DMEM&quot;)</code></pre>
</div>
</div>
<div id="overall-de-analysis-with-continuous-time-model" class="section level1">
<h1>Overall DE Analysis with Continuous time model</h1>
<div id="basic-analysis" class="section level2">
<h2>Basic analysis</h2>
<p>First as per normal workflow we estimate the size factors and we estimate the dispersions.</p>
<pre class="r"><code>ddsHTSeq &lt;- estimateSizeFactors(ddsHTSeq)
ddsHTSeq &lt;- estimateDispersions(ddsHTSeq)</code></pre>
<p>Then as differential expression test we choose a likelihood ratio test, where the interaction between time and condition is our main model, and the baseline we want to test against is a linear model without condition and with time as linear factor without interactions between time and condition.</p>
<pre class="r"><code>ddsHTSeq &lt;- nbinomLRT(ddsHTSeq, full = ~condition + condition:spline1 + condition:spline2 + 
    condition:spline3 + condition:spline4, reduced = ~spline1 + spline2 + spline3 + 
    spline4)</code></pre>
<p>If we visualise the results description</p>
<pre class="r"><code>res &lt;- results(ddsHTSeq)
res@elementMetadata@listData$description[2]
res@elementMetadata@listData$description[5]</code></pre>
<p>We will notice that p-values are calculated by contrasting the model</p>
<p><span class="math inline">\(y_{i} = condition_{i} + condition:spline1_{i} + condition:spline2_{i} + condition:spline3_{i} + condition:spline4_{i}\)</span></p>
<p>versus the base model</p>
<p><span class="math inline">\(y_{i} = spline1_{i} + spline2_{i} + spline3_{i} + spline4_{i}\)</span></p>
<p>The <span class="math inline">\(log_{2}-Fold\)</span> change is expressed contrasting XVIVO with DMEM over time.</p>
<p>We can order the results:</p>
<pre class="r"><code>resOrdered &lt;- as.data.frame(res[order(res$pvalue), ])
resOrdered$gene &lt;- row.names(resOrdered)
pandoc.table(head(resOrdered, 10L)[c(2, 3, 6)])</code></pre>
<p>: (#tab:DESeqRes) DESeq2 Results. Top results of the DESeq2 Analysis with continuous time modelling.</p>
</div>
<div id="dispersion-and-fold-change-shrinkage" class="section level2">
<h2>Dispersion and Fold-change Shrinkage</h2>
<p>If we look at the dispersion plot, we will notice quite some dispersion, considering that limits have been set to +-4 fold-change.</p>
<pre class="r"><code>plotMA(res, ylim = c(-4, 4))</code></pre>
<p>We can try to shrinkage the fold changes for better ranking and visualisation.
In order to do that, we need to note down the name of the contrast used for the previous results, which is going to be the last one in the list (as reported in the fold change of the summary)</p>
<pre class="r"><code>resultsNames(ddsHTSeq)</code></pre>
<p>And we choose <em>conditionXVIVO.spline4</em></p>
<pre class="r"><code>resLFC &lt;- lfcShrink(ddsHTSeq, coef = &quot;conditionXVIVO.spline4&quot;, type = &quot;apeglm&quot;)</code></pre>
<p>This will give us a new dispersion scenario, as follows:</p>
<pre class="r"><code>plotMA(resLFC, ylim = c(-4, 4))</code></pre>
<p>And this seems to improve the reporting of the results and the ranking of genes.</p>
</div>
<div id="sample-pca-and-heatmap" class="section level2">
<h2>Sample PCA and Heatmap</h2>
<!--
#######################################################################
### EXTRACT THE COUNTS
#######################################################################
-->
<p>we can extract the normalised counts for use in other packages. In this case we choose the <strong>variance stabilised data</strong> because the standard deviation will be quite constant across the whole dynamic range.</p>
<pre class="r"><code>vsd &lt;- vst(ddsHTSeq, blind = FALSE)
normalisedCounts &lt;- assay(vsd)</code></pre>
<p>It is kind of interesting to run a PCA over the samples separating them by both condition and time point</p>
<pre class="r"><code>plotPCA(vsd, intgroup = c(&quot;condition&quot;, &quot;time&quot;))</code></pre>
<p>And we can plot a nested heatmap of the samples ordered by treatment (DMEM vs ex-vivo) and time point.</p>
<pre class="r"><code>select &lt;- order(rowMeans(counts(ddsHTSeq, normalized = TRUE)), decreasing = TRUE)[1:20]
df &lt;- as.data.frame(colData(ddsHTSeq)[, c(&quot;condition&quot;, &quot;time&quot;)])
pheatmap(normalisedCounts[select, ], cluster_rows = FALSE, show_rownames = FALSE, 
    cluster_cols = FALSE, annotation_col = df)</code></pre>
<p>The figure is interesting because we can clearly see that in some of the genes which do not appear to distinguish the samples overall, a differennce in the dynamic over time seems noticeable.</p>
</div>
</div>
<div id="time-course-gene-clustering" class="section level1">
<h1>Time-course Gene Clustering</h1>
<!--
######################################################################
## TC SEQ
######################################################################
-->
<p>With TCseq we need to create GRanges coordinates of the genes in the analysis</p>
<pre class="r"><code>genes &lt;- row.names(normalisedCounts)
genesUniverseCoordinates &lt;- readRDS(&quot;~/CODE/core/Rutilities/geneSymbols_coordinates_biomart.RDS&quot;)
## removing non standard chromosomes
genesUniverseCoordinates &lt;- genesUniverseCoordinates %&gt;% filter(!grepl(&quot;CHR&quot;, 
    chr))
### there will be several symbols in this list which are not in the database
### for some reasons
exprGenesCoordinates &lt;- data.frame(symbol = genes)
exprGenesCoordinates &lt;- exprGenesCoordinates %&gt;% left_join(genesUniverseCoordinates, 
    by = &quot;symbol&quot;)
## removing duplicated entries
exprGenesCoordinates = exprGenesCoordinates[!duplicated(exprGenesCoordinates$symbol), 
    ]
## fixing those not present in database with a dummy coordinate
exprGenesCoordinates[is.na(exprGenesCoordinates$chr), &quot;chr&quot;] &lt;- 1
exprGenesCoordinates[is.na(exprGenesCoordinates$start), &quot;start&quot;] &lt;- 0
exprGenesCoordinates[is.na(exprGenesCoordinates$end), &quot;end&quot;] &lt;- 1
## tcseq needs specific identifiers
names(exprGenesCoordinates) &lt;- c(&quot;id&quot;, &quot;chr&quot;, &quot;start&quot;, &quot;end&quot;)</code></pre>
<p>The samples annotation needs to have specific headers:</p>
<pre class="r"><code>samplesAnnotationTC &lt;- sampleTable[, c(1, 3, 4)]
names(samplesAnnotationTC) &lt;- c(&quot;sampleid&quot;, &quot;group&quot;, &quot;timepoint&quot;)</code></pre>
<p>And we need to use the non normalised counts unfortunately from DESeq2 because the package only accepts integers</p>
<pre class="r"><code>tca &lt;- TCA(design = samplesAnnotationTC, counts = counts(ddsHTSeq), genomicFeature = exprGenesCoordinates)</code></pre>
<p>Differential analysis can be run with the following command which is preliminary to identify gene clusters</p>
<pre class="r"><code>tca &lt;- DBanalysis(tca)</code></pre>
<div id="fuzzy-clustering" class="section level2">
<h2>Fuzzy clustering</h2>
<p>Once the basic calculations have been computed, we will skip any pairwise comparison that is not interesting for us, and go straight to the clustering.
First of all we need to generat ea time course table (which is rearranged by feature):</p>
<pre class="r"><code>tca &lt;- timecourseTable(tca, value = &quot;expression&quot;, norm.method = &quot;rpkm&quot;, filter = TRUE)</code></pre>
<p>Then a clustering can be run, with a chose method, and given the differences between the clustering methods we choose the fuzzy one which is shown to perform better on time course data.
The number of potential clusters should be estimated, and we indicate 10 in order to try and capture the maximum diversity of the sample.</p>
<pre class="r"><code>tca &lt;- timeclust(tca, algo = &quot;cm&quot;, k = 10, standardize = TRUE)</code></pre>
<p>and we can now plot the results. Since we used fuzzy cmeans clustering, data points are color-coded according to membership values.</p>
<pre class="r"><code>grid.newpage()
pushViewport(viewport(layout = grid.layout(5, 2)))
vplayout &lt;- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
print(clustplot[[1]], vp = vplayout(1, 1))
print(clustplot[[2]], vp = vplayout(1, 2))
print(clustplot[[3]], vp = vplayout(2, 1))
print(clustplot[[4]], vp = vplayout(2, 2))
print(clustplot[[5]], vp = vplayout(3, 1))
print(clustplot[[6]], vp = vplayout(3, 2))
print(clustplot[[7]], vp = vplayout(4, 1))
print(clustplot[[8]], vp = vplayout(4, 2))
print(clustplot[[9]], vp = vplayout(5, 1))
print(clustplot[[10]], vp = vplayout(5, 2))</code></pre>
</div>
<div id="de-genes-and-cluster-membership" class="section level2">
<h2>DE genes and cluster membership</h2>
<p>We can use the <em>clustResults</em> function to extract the memebership of the genes to different clusters, and combine them with the DE analysis performed above with DESeq2.</p>
<pre class="r"><code>clusters &lt;- data.frame(gene = names(clustResults(tca)@cluster), cluster = unname(clustResults(tca)@cluster))
clusteredDEresults &lt;- as.data.frame(resOrdered) %&gt;% left_join(clusters, by = &quot;gene&quot;)
write_tsv(clusteredDEresults, &quot;PXXX_results_DESeq2_spline_clusters.txt&quot;)</code></pre>
<p>If we filter the genes by adjusted pvalue we can then count the memberships of each cluster and see which cluster has the higher number of DE genes</p>
<pre class="r"><code>pandoc.table(clusteredDEresults %&gt;% filter(padj &lt; 0.001 &amp; abs(log2FoldChange) &gt; 
    2) %&gt;% group_by(cluster) %&gt;% summarise(sigGenes = n()))</code></pre>
<p>: (#tab:clusters) Cluster Membership. Number of significantly differentially expressed genes, belonging to each of the 10 identified clusters.</p>
<p>This highlights 2 elements:
* the highest number of DE genes in time-course analysis does not belong to any of the 10 clusters identified
* the second highest group belongs to cluster 3</p>
<pre class="r"><code>print(clustplot[[3]])</code></pre>
</div>
</div>
<div id="impulsede2-analysis" class="section level1">
<h1>ImpulseDE2 Analysis</h1>
<!--
######################################################################
## IMPULSE 2
######################################################################

-->
<p>We can now run instead an <em>ImpulseDE2</em> model, which includes both case/control design as well as transient regulation:</p>
<pre class="r"><code>## also impulse needs specific column names for the annotations
sampleAnnoationImpulse &lt;- samplesAnnotationTC
names(sampleAnnoationImpulse) &lt;- c(&quot;Sample&quot;, &quot;Condition&quot;, &quot;Time&quot;)
## Time needs to be numeric
sampleAnnoationImpulse$Time &lt;- as.numeric(as.character(sampleAnnoationImpulse$Time))
## Then case/control needs to be named as such (very picky...)  We set DMEM
## as control and XVIVO as cases
sampleAnnoationImpulse$Condition &lt;- ifelse(sampleAnnoationImpulse$Condition == 
    &quot;DMEM&quot;, &quot;control&quot;, &quot;case&quot;)
objectImpulseDE2 &lt;- runImpulseDE2(matCountData = counts(ddsHTSeq), dfAnnotation = sampleAnnoationImpulse, 
    boolCaseCtrl = TRUE, boolIdentifyTransients = TRUE, scaNProc = 1)</code></pre>
<p>The software has given us the following warning:
20558 out of 56814 genes do not have obserserved non-zero counts and are excluded</p>
<div id="impulse-de-results" class="section level2">
<h2>Impulse DE results</h2>
<pre class="r"><code>impulseResults &lt;- objectImpulseDE2$dfImpulseDE2Results[order(objectImpulseDE2$dfImpulseDE2Results$padj), 
    ]
write_tsv(impulseResults, &quot;PXXX_results_ImpulseDE2.txt&quot;)
pandoc.table(head(impulseResults, 10L)[c(4, 3, 17)])</code></pre>
<p>: (#tab:impulseRes) Top ImpulseDE2 Results. Most significant differentially expressed genes identified by the Impulse model.</p>
<p>These results are interesting because they show the genes with the highest overall pvalues are also genes with a monotonous trend over time.</p>
</div>
<div id="overlap-with-deseq2-analysis" class="section level2">
<h2>Overlap with DESeq2 Analysis</h2>
<p>If we select the genes with an omnibus pvalue &lt; <span class="math inline">\(1*10^{-4}\)</span> we can inspect the overlap between the 2 methods</p>
<pre class="r"><code>methodsOverlap &lt;- euler(list(A = unique(impulseResults$Gene[impulseResults$padj &lt; 
    1e-04]), B = unique(resOrdered$gene[resOrdered$padj &lt; 1e-04])))
plot(methodsOverlap, quantities = TRUE, legend = list(labels = c(&quot;ImpulseDE2&quot;, 
    &quot;DESeq2&quot;)))</code></pre>
</div>
<div id="monotonous-expression" class="section level2">
<h2>Monotonous expression</h2>
<p>And we can use this to print the trajectories of the top 10 most significant genes</p>
<pre class="r"><code>lsgplotsGenes &lt;- plotGenes(vecGeneIDs = NULL, scaNTopIDs = 10, objectImpulseDE2 = objectImpulseDE2, 
    boolCaseCtrl = TRUE, dirOut = NULL, strFileName = NULL, vecRefPval = NULL, 
    strNameRefMethod = NULL)</code></pre>
<pre class="r"><code>grid.newpage()
pushViewport(viewport(layout = grid.layout(5, 2)))
vplayout &lt;- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
print(lsgplotsGenes[[1]], vp = vplayout(1, 1))
print(lsgplotsGenes[[2]], vp = vplayout(1, 2))
print(lsgplotsGenes[[3]], vp = vplayout(2, 1))
print(lsgplotsGenes[[4]], vp = vplayout(2, 2))
print(lsgplotsGenes[[5]], vp = vplayout(3, 1))
print(lsgplotsGenes[[6]], vp = vplayout(3, 2))
print(lsgplotsGenes[[7]], vp = vplayout(4, 1))
print(lsgplotsGenes[[8]], vp = vplayout(4, 2))
print(lsgplotsGenes[[9]], vp = vplayout(5, 1))
print(lsgplotsGenes[[10]], vp = vplayout(5, 2))</code></pre>
</div>
<div id="transient-expression" class="section level2">
<h2>Transient expression</h2>
<p>We can have a look at those top significant with a transient expression:</p>
<pre class="r"><code>impulseTransient &lt;- impulseResults %&gt;% filter(isTransient == &quot;TRUE&quot;) %&gt;% arrange(padj, 
    impulseTOsigmoid_padj)
pandoc.table(head(impulseTransient))</code></pre>
<p>: (#tab:impulseTransient) Top Transient Genes. Most significant genes identified with a non-monotonous expression over time.</p>
<p>And plot their time course</p>
<pre class="r"><code>lsgplotsTransientGenes &lt;- plotGenes(vecGeneIDs = head(impulseTransient$Gene, 
    10L), objectImpulseDE2 = objectImpulseDE2, boolCaseCtrl = TRUE, dirOut = NULL, 
    strFileName = NULL, vecRefPval = NULL, strNameRefMethod = NULL)
grid.newpage()
pushViewport(viewport(layout = grid.layout(5, 2)))
print(lsgplotsTransientGenes[[1]], vp = vplayout(1, 1))
print(lsgplotsTransientGenes[[2]], vp = vplayout(1, 2))
print(lsgplotsTransientGenes[[3]], vp = vplayout(2, 1))
print(lsgplotsTransientGenes[[4]], vp = vplayout(2, 2))
print(lsgplotsTransientGenes[[5]], vp = vplayout(3, 1))
print(lsgplotsTransientGenes[[6]], vp = vplayout(3, 2))
print(lsgplotsTransientGenes[[7]], vp = vplayout(4, 1))
print(lsgplotsTransientGenes[[8]], vp = vplayout(4, 2))
print(lsgplotsTransientGenes[[9]], vp = vplayout(5, 1))
print(lsgplotsTransientGenes[[10]], vp = vplayout(5, 2))</code></pre>
</div>
</div>
<div id="enrichment-analyses" class="section level1">
<h1>Enrichment Analyses</h1>
<p>Create a universe from ENSEMBLE</p>
<pre class="r"><code>#### creating the Universe from ENSEMBL
universe &lt;- keys(org.Hs.egSYMBOL)
# all symbols
x &lt;- org.Hs.egSYMBOL
mapped_genes &lt;- mappedkeys(x)
universeSymbols &lt;- unlist(as.list(x[mapped_genes]))
### universe can also be created more appropriately from the analysis
universeFromAnalysis = unlist(mget(impulseResults$Gene[!is.na(impulseResults$Gene)], 
    org.Hs.egSYMBOL2EG, ifnotfound = NA))
### however, one must get rid of those symbols without easy correspondence to
### entrezIDs
universeFromAnalysisCleaned &lt;- universeFromAnalysis[which(universeFromAnalysis != 
    &quot;NA&quot;)]</code></pre>
<p>Then we can select the different lists we would like to test:</p>
<pre class="r"><code>sigImpulseAll &lt;- bitr(impulseResults %&gt;% filter(padj &lt; 1e-04) %&gt;% pull(Gene), 
    fromType = &quot;SYMBOL&quot;, toType = c(&quot;ENSEMBL&quot;, &quot;ENTREZID&quot;), OrgDb = org.Hs.eg.db)
sigImpulseTransient &lt;- bitr(impulseResults %&gt;% filter(padj &lt; 0.001) %&gt;% filter(isTransient == 
    &quot;TRUE&quot;) %&gt;% pull(Gene), fromType = &quot;SYMBOL&quot;, toType = c(&quot;ENSEMBL&quot;, &quot;ENTREZID&quot;), 
    OrgDb = org.Hs.eg.db)</code></pre>
<p>In order to run GSEA-like analyses we need a numeric vector of fold-changes to be associated to the gene identifiers (entrez): unfortunately <strong>ImpulseDE2</strong> does not output this fold changes because the model follow a different approach.
We can therefore get the fold changes from the normalised calculations in DESeq2 which are anyway used inside the Impulse algo anyway.</p>
<pre class="r"><code>sigImpulseAll &lt;- sigImpulseAll %&gt;% left_join(resOrdered %&gt;% dplyr::select(gene, 
    log2FoldChange), by = c(SYMBOL = &quot;gene&quot;)) %&gt;% filter(abs(log2FoldChange) &gt; 
    2)
sigImpulseAllList &lt;- sigImpulseAll$log2FoldChange
names(sigImpulseAllList) &lt;- sigImpulseAll$ENTREZID
sigImpulseAllList &lt;- sort(sigImpulseAllList, decreasing = TRUE)
sigImpulseTransient &lt;- sigImpulseTransient %&gt;% left_join(resOrdered %&gt;% dplyr::select(gene, 
    log2FoldChange), by = c(SYMBOL = &quot;gene&quot;))
sigImpulseTransientList &lt;- sigImpulseTransient$log2FoldChange
names(sigImpulseTransientList) &lt;- sigImpulseTransient$ENTREZID
sigImpulseTransientList &lt;- sort(sigImpulseTransientList, decreasing = TRUE)</code></pre>
<div id="time-course-de-genes" class="section level2">
<h2>Time-course DE genes</h2>
<pre class="r"><code>sigImpulseAllBP &lt;- enrichGO(gene = unique(sigImpulseAll$ENTREZID), OrgDb = &quot;org.Hs.eg.db&quot;, 
    ont = &quot;BP&quot;, universe = universeFromAnalysisCleaned, qvalueCutoff = 0.3, 
    readable = T)
# head(sigImpulseAllBP@result[c(1:7)], 50L) emapplot(sigImpulseAllBP)
sigImpulseAllBPX &lt;- setReadable(sigImpulseAllBP, &quot;org.Hs.eg.db&quot;, &quot;ENTREZID&quot;)
cnetplot(sigImpulseAllBPX)
enrichplot::upsetplot(sigImpulseAllBPX, 8)</code></pre>
</div>
<div id="transient-de-genes" class="section level2">
<h2>Transient DE genes</h2>
<pre class="r"><code>sigImpulseTransientegoBP &lt;- enrichGO(gene = unique(sigImpulseTransient$ENTREZID), 
    OrgDb = &quot;org.Hs.eg.db&quot;, ont = &quot;BP&quot;, universe = universeFromAnalysisCleaned, 
    qvalueCutoff = 0.3, readable = T)
head(sigImpulseTransientegoBP@result[c(1:7)], 50L)
emapplot(sigImpulseTransientegoBP)
sigImpulseTransientegoBPX &lt;- setReadable(sigImpulseTransientegoBP, &quot;org.Hs.eg.db&quot;, 
    &quot;ENTREZID&quot;)
cnetplot(sigImpulseTransientegoBPX)
enrichplot::upsetplot(sigImpulseTransientegoBP, 8)
sigImpulseTransientegoKEGG &lt;- enrichKEGG(gene = unique(sigImpulseTransient$ENTREZID), 
    organism = &quot;hsa&quot;, pvalueCutoff = 0.05)
sigImpulseTransientGSEA &lt;- gseGO(geneList = sigImpulseTransientList, OrgDb = org.Hs.eg.db, 
    ont = &quot;BP&quot;, nPerm = 10000, minGSSize = 10, maxGSSize = 1000, pvalueCutoff = 0.05)</code></pre>
<pre class="r"><code>save.image(&quot;PXXX_RNAseq_time-course_analysis_20190830.RData&quot;)</code></pre>
</div>
</div>
</div>

    
    

    

    

</main>

        <footer>

            <p class="copyright text-muted">&copy; All rights reserved. Powered by <a href="https://gohugo.io">Hugo</a> and <a href="https://github.com/calintat/minimal">Minimal</a></p>

        </footer>
       
    </body>

</html>

